text/3720-labeled-match.md

```diff
@@ -0,0 +1,1419 @@
+```
+
+The borrow checker already operates on basic blocks, and [can handle irreducible control flow](https://rust-lang.zulipchat.com/#narrow/channel/186049-t-types.2Fpolonius/topic/Borrow-checking.20irreducible.20control-flow.3F), so no specific changes are needed.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## irreducible control flow
+
+The `const continue` construct introduces a way of expressing [irreducible control flow](https://en.wikipedia.org/wiki/Control-flow_graph#Reducibility) in the rust surface language. As far as we know, there are no blockers (e.g. [borrow checking should be able to handle it](https://rust-lang.zulipchat.com/#narrow/channel/186049-t-types.2Fpolonius/topic/Borrow-checking.20irreducible.20control-flow.3F), but currently it is not specified that HIR to MIR desugaring can introduce irreducible control flow (this has been discussed in [#114047](https://github.com/rust-lang/rust/issues/114047)).
+
+So while there are no blockers for this particular RFC, once you have irreducible control flow in the language there is no way back.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Let's look at alternatives in turn
+
+## switch fall-through
+
+In C, the "feature" of `switch` blocks automatically falling through to the following branch is used to guarantee a direct jump between two states. This idea gives rise to examples like [Duff's device](https://en.wikipedia.org/wiki/Duff%27s_device):
+
+```c
+send(to, from, count)
+register short *to, *from;
+register count;
+{
+    register n = (count + 7) / 8;
+    switch (count % 8) {
+    case 0: do { *to = *from++;
+    case 7:      *to = *from++;
+    case 6:      *to = *from++;
+    case 5:      *to = *from++;
+    case 4:      *to = *from++;
+    case 3:      *to = *from++;
+    case 2:      *to = *from++;
+    case 1:      *to = *from++;
+            } while (--n > 0);
+    }
+}
+```
+
+This fall-through behavior is often considered unintuitive, to the point that in many C code bases, such fall-throughs are explicitly labeled with a comment to call attention to the fact that the fall-through is deliberate. But this feature is a part of C for a reason: the fall-through is an unconditional jump, which is often essential for good performance.
+
+The `loop match` proposal has two major advantages over fallthrough:
+
+- there is no need to list branches in a particular order
+- more than one next state can be reached with a direct jump
+
+It turns out that `loop match` is fairly expressive, and can in fact express Duff's device:
+
+```rust
+// originally written by Ralf Jung on zullip
+// assumes count > 0
+// `one()` performs a one-byte write and increments the counters
+let mut n = count.div_ceil(4);
+loop match count % 4 {
+  0 => { one(); const continue 3 }
+  3 => { one(); const continue 2 }
+  2 => { one(); const continue 1 }
+  1 => { one(); n -= 1; if n > 0 { continue count % 4; } }
```

Comment:
Also I'm pretty sure the loop is wrong (obviously no exit condition, and less obviously you should not restart from `count % 4`). I think it suggests the proposed syntax will still lead to subtle bugs when the programmer is not careful enough.

```suggestion
  1 => { one(); n -= 1; if n > 0 { const continue 0; } else { break; } }
```

Reply:
I think the syntax is not really at fault here, rather that it's just tricky control flow.